### ISSUE TITLE
Taxi.hs issues

### ISSUE THREAD

**Issue by jmtd** [striot/striot#5]
Using the following input

```
-rw-r--r-- 1 jon jon 367M Feb 28 11:28 "sorteddata.csv"
b57e16b390d81036a87cf9cf29e8ff276959e1d6  sorteddata.csv
```

With current master (30b47df72e7688be522d503201cd09058c60843b) I get the following output

```
λ> mainQ2
*** Exception: Map.!: given key is not an element in the map
CallStack (from HasCallStack):
  error, called at libraries/containers/Data/Map/Base.hs:489:16 in containers-0.5.7.1:Data.Map.Base
*Taxi
```

Winding back in time to ec30d7b27ac5295c30af9fa199a7c45adf247592 and attempting it there, I got a segfault from GHCi!

```
λ> mainQ2                                                            
Segmentation fault  
▶ ghci --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1
```

**Comment by jmtd** [comment]
The various other main* functions work fine, both historically and with 30b47df.

**Comment by PaulWatson9994** [comment]
I've dug around in taxi.hs and made some progress. I redefined profitability so that it works without having to have every cell already included in the 2 maps it takes as input (by using Map.findWithDefault instead of Map.!):

`profitability emptyTaxis cellProf = foldl (\m c->Map.insert c (Map.findWithDefault 0 c cellProf / fromIntegral (Map.findWithDefault 0 c emptyTaxis)) m) Map.empty (Map.keys emptyTaxis)
`

Also (for efficiency), it only calculates results for the subset of cells that are included in the emptyTaxis map: that's the `Map.keys emptytaxis` part.

Q2 then works on small numbers of trips without the error I was getting before:

e.g. if we define:

q2TripSourceTest9 trips = do contents <- readFile "sorteddata.csv"
                             putStr $ show $ profitableCells $ take trips $ tripSource contents  

then call q2TripSourceTest9 50
It gives a result.

I need to let run mainQ2 for longer to check that gives a result eventually (I killed it after a few minutes); however, it may be that the top-k implementation I used in the streamWindowAggregate part of profitableCells is very slow when dealing with the full 130900 trips in contents.csv - something to look at (anyone know a good Haskell top-k implementation?)
